streams/src/test/java/org/apache/kafka/streams/integration/PurgeRepartitionTopicIntegrationTest.java

```diff
@@ -0,0 +1,222 @@
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Properties;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+
+@Category({IntegrationTest.class})
+public class PurgeRepartitionTopicIntegrationTest {
+
+    private static final int NUM_BROKERS = 1;
+
+    private static final String INPUT_TOPIC = "input-stream";
+    private static final String APPLICATION_ID = "restore-test";
+    private static final String REPARTITION_TOPIC = APPLICATION_ID + "-KSTREAM-AGGREGATE-STATE-STORE-0000000002-repartition";
+
+    private static AdminClient adminClient;
+    private static KafkaStreams kafkaStreams;
+    private static Integer purgeIntervalMs = 10;
+    private static Integer purgeSegmentBytes = 2000;
+
+    @ClassRule
+    public static final EmbeddedKafkaCluster CLUSTER = new EmbeddedKafkaCluster(NUM_BROKERS, new Properties() {
+        {
+            put("log.retention.check.interval.ms", purgeIntervalMs);
+            put(TopicConfig.FILE_DELETE_DELAY_MS_CONFIG, 0);
+        }
+    });
+
+    private Time time = CLUSTER.time;
+
+    private class RepartitionTopicCreatedWithExpectedConfigs implements TestCondition {
+        @Override
+        final public boolean conditionMet() {
+            try {
+                Set<String> topics = adminClient.listTopics().names().get();
+
+                if (!topics.contains(REPARTITION_TOPIC)) {
+                    return false;
+                }
+            } catch (final Exception e) {
+                return false;
+            }
+
+            try {
+                ConfigResource resource = new ConfigResource(ConfigResource.Type.TOPIC, REPARTITION_TOPIC);
+                Config config = adminClient.describeConfigs(Collections.singleton(resource))
+                        .values().get(resource).get();
+                return config.get(TopicConfig.CLEANUP_POLICY_CONFIG).value().equals(TopicConfig.CLEANUP_POLICY_DELETE)
+                        && config.get(TopicConfig.SEGMENT_MS_CONFIG).value().equals(purgeIntervalMs.toString())
+                        && config.get(TopicConfig.SEGMENT_BYTES_CONFIG).value().equals(purgeSegmentBytes.toString());
+            } catch (final Exception e) {
+                return false;
+            }
+        }
+    }
+
+    private interface TopicSizeVerifier {
+
+        boolean verify(long currentSize);
+    }
+
+    private class RepartitionTopicVerified implements TestCondition {
+        private final TopicSizeVerifier verifier;
+
+        RepartitionTopicVerified(TopicSizeVerifier verifier) {
+            this.verifier = verifier;
+        }
+
+        @Override
+        public final boolean conditionMet() {
+            time.sleep(purgeIntervalMs);
```

Comment:
But we advance the time before we verify that the topic was written. I think, this is ok for the "purging" test but not for the first test for which we expect to see data in the repartitioning topic. As the StreamsThreads run in the back ground, in the first check directly after we advanced time, we could have a context switch and data might get purged and thus, we can never hit the condition that "repartitionsTopicSize > 0"

Or can this never happen?

Reply:
This should never happen, since starting offset based purging should never be "clean". I.e. even if you set the purge offset to be LEO it should not delete the topic partition segments to empty.

Reply:
Ack. Thx.